docs/csharp/tutorials/exploration/patterns-objects.md

```diff
@@ -0,0 +1,124 @@
+
+## Implement the commands with patterns
+
+A better way is to use *patterns* to determine if the object is in a valid state to execute a command. You can express if a command is allowed as a function of three variables: the state of the gate, the level of the water, and the new setting:
+
+| new setting | Gate state | Water Level | result             |
+| ----------- | ---------- | ----------- | ------------------ |
+| Closed      | Closed     | High        | Closed             |
+| Closed      | Closed     | Low         | Closed             |
+| Closed      | Open       | High        | Open               |
+| ~~Closed~~  | ~~Open~~   | ~~Low~~     | ~~Closed~~         |
+| Open        | Closed     | High        | Open               |
+| Open        | Closed     | Low         | Closed (Error)     |
+| Open        | Open       | High        | Open               |
+| ~~Open~~    | ~~Open~~   | ~~Low~~     | ~~Closed (Error)~~ |
+
+The fourth and last rows in the table have strike through text because they're invalid. The code you're adding now should make sure the high water gate is never opened when the water is low.  Those states can be coded as a single switch expression (remember that `false` indicates "Closed"):
+
+:::code language="csharp" source="snippets/pattern-objects/InterimSteps.cs" ID="ThirdImplementation":::
+
+Try this version. Your tests pass, validating the code. The full table shows the possible combinations of inputs and results. That means you and other developers can quickly look at the table and see that you've covered all the possible inputs. Even easier, the compiler can help as well. After you add the previous code, you can see that the compiler generates a warning: *CS8524* indicates the switch expression doesn't cover all possible inputs. The reason for that warning is that one of the inputs is an `enum` type. The compiler interprets "all possible inputs" as all inputs from the underlying type, typically an `int`. This `switch` expression only checks the values declared in the `enum`. To remove the warning, you can add a catch-all discard pattern for the last arm of the expression. This condition throws an exception, because it indicates invalid input:
+
+```csharp
+_  => throw new InvalidOperationException("Invalid internal state"),
+```
+
+The preceding switch arm must be last in your `switch` expression because it matches all inputs. Experiment by moving it earlier in the order. That causes a compiler error *8510* for unreachable code in a pattern.  The natural structure of switch expressions enables the compiler to generate errors and warnings for possible mistakes. That makes it easier for you to create correct code in fewer iterations. That safety gives you the freedom to combine switch arms with wildcards. The compiler will issue errors if your combination results in unreachable arms you didn't expect. The compiler issues warnings if you remove an arm that's needed.
```

Comment:
```suggestion
The preceding switch arm must be last in your `switch` expression because it matches all inputs. Experiment by moving it earlier in the order. That causes a compiler error *CS8510* for unreachable code in a pattern.  The natural structure of switch expressions enables the compiler to generate errors and warnings for possible mistakes. The compiler "safety net" makes it easier for you to create correct code in fewer iterations, and the freedom to combine switch arms with wildcards. The compiler will issue errors if your combination results in unreachable arms you didn't expect, and warnings if you remove an arm that's needed.
```

It seemed like there were too many sentences that started with "that".

Reply:
great suggestion.